### ISSUE TITLE
12/17/24 - TO DO AFTER THE BREAK

### ISSUE THREAD

**Issue by badeaa3** [badeaa3/asic-testing-fnal-lab-notebooks#46]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 55
> Originally opened by bparpillon on 2024-12-17T15:12:47Z

### CODE AND DNN OPTIMIZATION

- [ ] hard code CASES for different frequencies that automatically set all the delay values required to work instead of passing 4-5 different arguments every time 
- [ ] dnn_csv load takes 0.2 to 0.4 sec and should be optimized : maybe save the csv in the SDcard
- [ ] shift register programming requires more than 10,000 writes operations, this could be optimized by creating a stream_set_memory command
- [ ] Incorporate the shift register routine within the DNN routine
- [ ] scanchain readback on DATA_ARRAY_0 can be optimized by NOT extracting the y-profile thus saving 24 read instructions. The argument **readYproj** should be set to **False**
- [ ] scanchain readback on DATA_ARRAY_1 can also be optimized by NOT reading the BxCLK data. A FW upgrade would be necessary to optimize the dnn_output data and compress the essential bits in a single word.

### TRAINING OPTIMIZATION
- [ ] Verify the DNN extraction. We should not have any 3's coming from the chip
- [ ] Simplify the DNN evaluation flow
- [ ] Run the training without the asic included. Verify that we can get weights with the expected 75% accuracy
- [ ] Feature space to explore
        1. alpha, start low and then progressively turn up
        2. send each pattern 10 times to **guide** the loss optimization against noise 
        3. Enlarge the training and separate from eval set
 
### RADIATION TRAINING
- [ ] Force ileak = [0 - 50nA]
- [ ] Extract S-curve and show impact
- [ ] Retrain DNN

### **ANALOG**
- [ ] Extract Scurve on all pixel on many dies for both superpixel
- [ ] Estimate Charge dispersion

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 55
> Originally posted by bparpillon on 2024-12-18T14:40:27Z

**To extract the new weight/csv file:**
1. update this line with the model file https://github.com/smart-pix/CMSPIX28Spacely/blob/main/CMSPIX28Spacely_Subroutines_B5_DNNTraining.py#L417
2. uncomment it and comment out the previous line
3. set this to false https://github.com/smart-pix/CMSPIX28Spacely/blob/main/CMSPIX28Spacely_Subroutines_B5_DNNTraining.py#L438
4. then run the file `python CMSPIX28Spacely_Subroutines_B5_DNNTraining.py`

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 55
> Originally posted by badeaa3 on 2024-12-26T06:41:45Z

Readings to brainstorm ideas:
- [HAQ: Hardware-Aware Automated Quantization with Mixed Precision](https://arxiv.org/abs/1811.08886)
- [Agentic AI](https://blogs.nvidia.com/blog/what-is-agentic-ai/)
- Attempt to train model with pytorch and brevitas then send weights to ASIC. https://xilinx.github.io/brevitas/getting_started
    - Attempt to send pytorch model through hls4ml
- Explore using [Liquid Time-constant Networks](https://arxiv.org/abs/2006.04439) to handle continuous time series of pixel data
